src/rgw/rgw_lua_utils.h

```diff
@@ -233,12 +348,67 @@ int StringMapWriteableNewIndex(lua_State* L) {
       map->insert_or_assign(index, value);
     }
   } else {
-    map->erase(std::string(index));
+    // erase the element. since in lua: "t[index] = nil" is removing the entry at "t[index]"
+    if (const auto it = map->find(index); it != map->end()) {
+      // index was found
+      update_erased_iterator<MapType>(L, it, map->erase(it));
+    }
   }
 
   return NO_RETURNVAL;
 }
 
+// implements the lua next() function for iterating over a table
+// first argument is a table and the second argument is an index in this table
+// returns the next index of the table and its associated value
+// when input index is nil, the function returns the initial value and index
+// when the it reaches the last entry of the table it return nil as the index and value
+template<typename MapType, typename ValueMetaType=void>
+int next(lua_State* L) {
+  using Iterator = typename MapType::iterator;
+  auto map = reinterpret_cast<MapType*>(lua_touserdata(L, lua_upvalueindex(FIRST_UPVAL)));
+  ceph_assert(map);
+  Iterator* next_it = nullptr;
+
+  if (lua_isnil(L, 2)) {
+    // pop the 2 nils
+    lua_pop(L, 2);
+    // create userdata
+    next_it = create_iterator_metadata<MapType>(L, map->begin());
+  } else {
+    next_it = reinterpret_cast<Iterator*>(lua_touserdata(L, 2));
+    ceph_assert(next_it);
+    *next_it = std::next(*next_it);
+  }
+
+  if (*next_it == map->end()) {
+    // index of the last element was provided
+    destroy_iterator<MapType>(L);
```

Comment:
this doesn't look reliable in the case of `break` statements, but maybe it's the best we can do?

Reply:
now added to the `__gc` callback of the metatable

Reply:
repushed a simpler implementation of the `__gc` callback in: https://github.com/ceph/ceph/commit/45d233535ad585f72fb5218e452131d225206afd